Sources/Commands/PinsStore.swift

```diff
@@ -0,0 +1,188 @@
+/*
+ This source file is part of the Swift.org open source project
+
+ Copyright 2015 - 2016 Apple Inc. and the Swift project authors
+ Licensed under Apache License v2.0 with Runtime Library Exception
+
+ See http://swift.org/LICENSE.txt for license information
+ See http://swift.org/CONTRIBUTORS.txt for Swift project authors
+*/
+
+import Basic
+import SourceControl
+import Utility
+
+public enum PinOperationError: Swift.Error {
+    case alreadyPinned
+    case notPinned
+}
+
+public struct PinsStore {
+    public struct Pin {
+        /// The repository specifer of the pinned dependency.
+        public let repository: RepositorySpecifier
+
+        /// The pinned version.
+        public let version: Version
+
+        /// The reason text for pinning this dependency.
+        public let reason: String?
+
+        init(repository: RepositorySpecifier, version: Version, reason: String? = nil) {
+            self.repository = repository
+            self.version = version
+            self.reason = reason
+        }
+    }
+
+    /// The path to the pins file.
+    fileprivate let pinsFile: AbsolutePath
+
+    /// The filesystem to manage the pin file on.
+    fileprivate var fileSystem: FileSystem
+
+    /// The pins map.
+    fileprivate(set) var pinsMap: [RepositorySpecifier: Pin]
+
+    /// The current pins.
+    public var pins: AnySequence<Pin> {
+        return AnySequence<Pin>(pinsMap.values)
+    }
+
+    /// Create a new pins store.
+    ///
+    /// - Parameters:
+    ///   - pinsFile: Path to the pins file.
+    ///   - fileSystem: The filesystem to manage the pin file on.
+    public init(pinsFile: AbsolutePath, fileSystem: FileSystem) throws {
+        self.pinsFile = pinsFile
+        self.fileSystem = fileSystem
+        pinsMap = [:]
+        try restoreState()
+    }
+
+    /// Pin a repository at a version.
+    ///
+    /// - Parameters:
+    ///   - repository: The repository to pin. It should not be already pinned.
```

Comment:
Why should it not already be pinned? From an API perspective, it doesn't seem to matter.